asset-transfer-private-data/chaincode-go/go/privateAssetTransfer.go

```diff
@@ -0,0 +1,503 @@
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal JSON: %s", err.Error())
+	}
+
+	// delete the asset from state
+	err = ctx.GetStub().DelPrivateData("assetCollection", assetDeleteInput.ID)
+	if err != nil {
+		return fmt.Errorf("Failed to delete state:" + err.Error())
+	}
+
+	// Also delete the asset from the color~ID index
+	indexID := "color~ID"
+	colorIDIndexKey, err := ctx.GetStub().CreateCompositeKey(indexID, []string{assetToDelete.Color, assetToDelete.ID})
+	if err != nil {
+			return err
+	}
+	err = ctx.GetStub().DelPrivateData("assetCollection", colorIDIndexKey)
+	if err != nil {
+		return fmt.Errorf("Failed to delete asset:" + err.Error())
+	}
+
+	// Finally, delete private details of asset
+
+	collection, err := getCollectionName(ctx) // get owners collection
+
+	err = ctx.GetStub().DelPrivateData(collection, assetDeleteInput.ID) // delete the asset
+	if err != nil {
+			return err
+	}
+
+	return nil
+
+}
+
+// Delete agreement to buy asset
+func (s *SmartContract) DeleteAgreement(ctx contractapi.TransactionContextInterface) error {
+
+	transMap, err := ctx.GetStub().GetTransient()
+		if err != nil {
+			return fmt.Errorf("Error getting transient: " + err.Error())
+		}
+
+	// Asset properties are private, therefore they get passed in transient field
+	transientDeleteJSON, ok := transMap["agree_delete"]
+		if !ok {
+			return fmt.Errorf("asset to delete not found in the transient map")
+		}
+
+	type assetDelete struct {
+			ID string `json:"ID"`
+		}
+
+	var assetDeleteInput assetDelete
+	err = json.Unmarshal(transientDeleteJSON, &assetDeleteInput)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal JSON: %s", err.Error())
+		}
+
+	if len(assetDeleteInput.ID) == 0 {
+		return fmt.Errorf("ID field must be a non-empty string")
+	}
+
+	// delete private details of agreement
+
+	collection, err := getCollectionName(ctx) // get owners collection
+
+	err = ctx.GetStub().DelPrivateData(collection, assetDeleteInput.ID) // delete the asset
+	if err != nil {
+			return err
+	}
+
+	// delete agreement record
+
+	assetAgreeKey, err := ctx.GetStub().CreateCompositeKey("agreeToPurchase",[]string{assetDeleteInput.ID}) // create composite key
+		if err != nil {
+			return fmt.Errorf("failed to create composite key: %s", err.Error())
+		}
+
+	err = ctx.GetStub().DelState(assetAgreeKey) // remove agreement from state
+		if err != nil {
+			return err
+		}
+
+	return nil
+
+}
+
+// Return the ID of submitting client identity. This will be used to identify
+// and verify the owner on chain
+func (s *SmartContract) ReturnID(ctx contractapi.TransactionContextInterface) (string, error) {
+
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return "", fmt.Errorf("failed to get verified OrgID: %s", err.Error())
+	}
+	return clientID, nil
+}
+
+// helper function to get collection of submitting client identity
+func getCollectionName(ctx contractapi.TransactionContextInterface) (string, error) {
```

Comment:
I think we'll want to do something similar to:
https://github.com/hyperledger/fabric-samples/blob/master/asset-transfer-secured-agreement/chaincode-go/asset_transfer.go#L473-L476

That would ensure that Org1 client can't attempt to create asset on Org2 peer (it would fail at endorsement time, instead of validation time due to endorsement failure, which is easier to understand).

Reply:
Going to do this within the `getCollectionName` function to keep things concise in this simpler smart contract.

```
func getCollectionName(ctx contractapi.TransactionContextInterface) (string, error) {

	// Get the MSP ID of submitting client identity
	clientMSPID, err := ctx.GetClientIdentity().GetMSPID()
	if err != nil {
		return "", fmt.Errorf("failed to get verified OrgID: %s", err.Error())
	}

	// verify that the client is submitting request to peer in their organization
	peerOrgID, err := shim.GetMSPID()
	if err != nil {
		return fmt.Errorf("failed getting peer's orgID: %s", err.Error())
	}

	if clientMSPID != peerOrgID {
		return fmt.Errorf("client from org %s is not authorized to read or write private data on org %s peer", clientMSPID, peerOrgID)
	}

	// create the collection name
	orgCollection := clientMSPID + "DetailsCollection"

	return orgCollection, nil
}
```